### PR TITLE
feat(node): implement api config validation for cookie-based auth

### PR DESCRIPTION
**Pull request by jensneuse**
This PR adds API config validation which detects wrong config for cookie-based auth configs.

### CODE REVIEW

File: packages/sdk/CHANGELOG.md [hunk 1]
```diff
@@ -7,24 +7,24 @@ See [Conventional Commits](https://conventionalcommits.org) for commit guideline
 
 ### Features
 
-* implement userId for fromClaim directive ([#152](https://github.com/wundergraph/wundergraph/issues/152)) ([51df6e5](https://github.com/wundergraph/wundergraph/commit/51df6e50244bee9f5f8d579ff6f604e1a1c853d9)) (@jensneuse)
```
Comment (StarpTech) [inline packages/sdk/CHANGELOG.md:10]: Why is it changed? This is auto generated.

File: pkg/node/node.go [hunk 2]
```diff
@@ -308,6 +310,34 @@ func (n *Node) startServer(nodeConfig wgpb.WunderNodeConfig) {
...
+			// in dev mode, ignore missing cookie secrets
+			// we automatically override them
+			IgnoreMissingCookieAuthenticationSecrets: n.options.devMode,
+		})
+
+		if !valid {
+			n.log.Error("API config invalid",
+				abstractlogger.Strings("errors", messages),
+			)
+			os.Exit(1)
```
Comment (StarpTech) [inline pkg/node/node.go:323]: or use `log.fatal`? I think this ensures that the log buffer is printed before the process dies.

File: pkg/node/node.go [hunk 3]
```diff
@@ -308,6 +310,34 @@ func (n *Node) startServer(nodeConfig wgpb.WunderNodeConfig) {
...
+			return
+		}
+
+		n.setApiConfigDefaults(api)
+
+		// revalidate after setting defaults
+		// just to make sure that defaults are validated
+		valid, messages = validate.DefaultApiConfig(api, validate.ApiConfigValidationConfig{})
```
Comment (StarpTech) [inline pkg/node/node.go:331]: Why do we need to call the validator twice? This is confusing to me.

File: pkg/validate/apiconfig.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+type ApiConfigValidationConfig struct {
+	IgnoreMissingCookieAuthenticationSecrets bool
+}
+
+type ApiConfigValidationRule interface {
+	Validate(a *ApiConfigValidator)
+}
+
+type ApiConfigValidator struct {
+	api      *wgpb.Api
```
Comment (StarpTech) [inline pkg/validate/apiconfig.go:36]: Is it really necessary that the validator has to know the `*wgpb.Api` object? The validator only needs to know: What is a valid value for a cookie secret / csrf.

```go
err := Validate(a.api.AuthenticationConfig.CookieBased.CsrfSecret, Validate.CookieCSRF)
err := Validate(a.api.AuthenticationConfig.CookieBased.BlockKey, Validate.CookieSecret)
```

I think this is easier to write and maintain.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: packages/sdk/CHANGELOG.md (hunk 1 above)
Comment (jensneuse) [reply packages/sdk/CHANGELOG.md:10]: I don't know

File: pkg/node/node.go (hunk 3 above)
Comment (jensneuse) [reply pkg/node/node.go:331]: the comment says it

File: pkg/validate/apiconfig.go (hunk 4 above)
Comment (jensneuse) [reply pkg/validate/apiconfig.go:36]: subjective

Comment (jensneuse) [reply pkg/validate/apiconfig.go:36]: I've investigated a few other solutions. The issue remains that somewhere needs to sit the mapping.
I'll abandon the idea of improving the code at this point.
If you'd like to refactor it, ok.

File: packages/sdk/CHANGELOG.md (hunk 1 above)
Comment (jensneuse) [reply packages/sdk/CHANGELOG.md]: fixed

File: pkg/validate/apiconfig.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/wundergraph/wundergraph/pkg/apiconfig"
+	"github.com/wundergraph/wundergraph/pkg/loadvariable"
+	"github.com/wundergraph/wundergraph/types/go/wgpb"
+)
+
+// ApiConfig validates the Api
+func ApiConfig(api *wgpb.Api) (valid bool, messages []string) {
+
+	if !apiconfig.HasCookieAuthEnabled(api) {
+		// skip following tests if cookie auth is not enabled
```
Comment (StarpTech) [inline pkg/validate/apiconfig.go:13]: If you put this here, we can only validate `AuthenticationConfig` settings, but `ApiConfig` is a place to validate the whole config.

Comment (jensneuse) [reply pkg/validate/apiconfig.go:13]: For each rule, we should call one function which can return early. Currently, with just one rule, it's fine.

Comment (StarpTech) [reply pkg/validate/apiconfig.go:13]: ok

**Pull request merged**